Bluetooth: Controller: Adds path loss monitoring functionality

tyler-huffman-demant opened the pull request:
This functionality is implemented in the HCI-facing side of the controller and the ULL. LLL functionality is not implemented in this commit.

Furthermore, this removes a dependency for LE Power Control, as the features are related, but not necessarily required.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_conn.h:
```diff
@@ -14,6 +14,27 @@
+
+struct path_loss_params {
+	uint16_t min_time_spent;
+	uint8_t  enabled;
```
Some values like `enabled` and `conn_handle` can use fewer bits than the full size of their types. Consider if we can optimize the size here a bit

tyler-huffman-demant replied:
In practice, I think they can be optimized (like `enabled`), though it seems the Bluetooth spec only specifies between 1 and 2 octets without specifying meaningful bits, so I'd prefer to keep it as is.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2910,6 +2910,56 @@ uint8_t ull_conn_lll_phy_active(struct ll_conn *conn, uint8_t phys)
+
+	conn = ll_connected_get(handle);
+
+	if (!conn) {
```
```suggestion
	if (conn == NULL) {
```

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/92450#discussion_r2185255112

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2910,6 +2910,56 @@ uint8_t ull_conn_lll_phy_active(struct ll_conn *conn, uint8_t phys)
+
+	conn = ll_connected_get(handle);
+
+	if (!conn) {
```
```suggestion
	if (conn == NULL) {
```

tyler-huffman-demant commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8645,6 +8700,28 @@ static void le_req_peer_sca_complete(struct pdu_data *pdu, uint16_t handle,
+	struct bt_hci_evt_le_path_loss_threshold *sep;
+	struct node_rx_path_loss *pl_pdu;
+
+	pl_pdu = (void *)pdu_data;
```
AFAIK `struct pdu_data` is the PDU's payload, so long as the raw PDU is correctly following the Bluetooth spec, then it shouldn't be disjoint.

But more importantly, I just modeled it after these functions:

https://github.com/zephyrproject-rtos/zephyr/blob/fdd02c5105d2a40cf6410cf89ccd22c43a48399e/subsys/bluetooth/controller/hci/hci.c#L8584

https://github.com/zephyrproject-rtos/zephyr/blob/fdd02c5105d2a40cf6410cf89ccd22c43a48399e/subsys/bluetooth/controller/hci/hci.c#L8606

https://github.com/zephyrproject-rtos/zephyr/blob/fdd02c5105d2a40cf6410cf89ccd22c43a48399e/subsys/bluetooth/controller/hci/hci.c#L8631

Thalley replied:
Hmm, yeah, this is outside of my knowledge. 

I was just looking an comparing the 2 structs, and they don't really share any common fields, so that's why casting one to the other seems weird. Someone who knows this better than I should comment here :)

tyler-huffman-demant commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -343,3 +343,12 @@ void ll_coex_ticker_id_get(uint8_t * const instance_index,
+				     uint8_t  high_hysteresis,
+				     uint8_t  low_threshold,
+				     uint8_t  low_hysteresis,
+				     uint16_t min_time_spent);
```
Fixed 👍

tyler-huffman-demant commented on subsys/bluetooth/controller/ll_sw/lll_conn.h:
```diff
@@ -14,6 +14,27 @@
+struct path_loss_event {
+	uint8_t curr_path_loss;
+	uint8_t zone_entered;
+};
```
Exactly right; good catch 👍

tyler-huffman-demant commented on subsys/bluetooth/controller/ll_sw/lll_conn.h:
```diff
@@ -14,6 +14,27 @@
+	uint16_t conn_handle;
+	uint16_t min_time_counter;
+	uint8_t  new_zone;
+};
```
I don't see why not 👍 😄

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2793,6 +2793,50 @@ static void le_set_phy(struct net_buf *buf, struct net_buf **evt)
+	struct bt_hci_cp_le_set_path_loss_reporting_parameters *cmd = (void *)buf->data;
+	struct bt_hci_rp_le_set_path_loss_reporting_parameters *rp;
+
+	uint16_t handle = sys_le16_to_cpu(cmd->handle);
```
Sure, could be a nice first-time issue as it's just fixing a missing check several places :) Chances are that it's being handled somewhere else that I'm not aware of @cvinayak ?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8645,6 +8700,28 @@ static void le_req_peer_sca_complete(struct pdu_data *pdu, uint16_t handle,
+
+	pl_pdu = (void *)pdu_data;
+
+	if (!(event_mask & BT_EVT_MASK_LE_META_EVENT) ||
```
I'll defer that to the maintainer of the controller. The reason why I mentioned here is that it's in violation of rule 85 of the Zephyr coding guidelines that does not allow using integers as controlling expression https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html : https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8645,6 +8700,28 @@ static void le_req_peer_sca_complete(struct pdu_data *pdu, uint16_t handle,
+	pl_pdu = (void *)pdu_data;
+
+	if (!(event_mask & BT_EVT_MASK_LE_META_EVENT) ||
+	    !(le_event_mask & BT_EVT_MASK_LE_PATH_LOSS_THRESHOLD)) {
```
See https://github.com/zephyrproject-rtos/zephyr/pull/92450#discussion_r2185255112

tyler-huffman-demant committed:
Bluetooth: Controller: Adds path loss monitoring functionality

This functionality is implemented in the HCI-facing side of the
controller and the ULL. LLL functionality is not implemented in this
commit.

Furthermore, this removes a dependency for LE Power Control, as the
features are related, but not necessarily required.

Signed-off-by: Tyler Joseph Huffman <tyhu@demant.com>

tyler-huffman-demant commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2793,6 +2793,50 @@ static void le_set_phy(struct net_buf *buf, struct net_buf **evt)
+	rp->handle = handle;
+
+	/* Anything beyond 0x00 or 0x01 is reserved */
+	if (cmd->enable > 0x01) {
```
Done 👍

Thalley reviewed (approved):
No blocking comments from my side

cvinayak reviewed (approved):
QUICK-ACK. No objections.

tyler-huffman-demant commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -720,6 +720,11 @@ struct bt_hci_cp_le_set_path_loss_reporting_enable {
+	uint8_t  status;
+	uint16_t handle;
+} __packed;
+
```
@jhedberg I made the changes. Could you take a look?
